### PR TITLE
Dagstermill tutorial

### PR DESCRIPTION
**Pull request by puma314**
Added tutorial for dagstermill under dagstermill/docs/intro_tutorial.md

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/infos.py [hunk 1]
```diff
@@ -68,17 +68,21 @@ class TransformExecutionInfo(
...
-            check.inst_param(solid, 'solid', Solid),
+            check.opt_inst_param(step, 'step', ExecutionStep),
+            # step can be None if TransformExecutionInfo is being reconstructed
+            # out-of-pipeline execution
             check.inst_param(pipeline_def, 'pipeline_def', PipelineDefinition),
             context.resources,
             DagsterLog(context),
         )
 
     @property
```
Comment (schrockn) [inline python_modules/dagster/dagster/core/definitions/infos.py]: Let's add a @solid property as well to maintain API compatibility

File: python_modules/dagster/docs/guides/adding_event.md [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    if event_cls == StepMaterializationRecord:
+        step_args['file_name'] = logger_message.meta['file_name']
+        step_args['file_location'] = logger_message.meta['file_location']
+    ...
+```
+This is the end of all the changes necesary in the `events.py` file, but we still need to hook up emission of a Dagster `EventRecord` to a GraphQL schema and GraphQL event.
+
+`python_modules/dagit/dagit/schema/runs.py`
+
+The above file is where the function `from_dagster_event()` converts an `EventRecord` to an `Dauphin` object (which itself is a wrapper on Graphene objects). To add a Dauphin class, we actually create a wrapper class that uses this `Meta` incantation that auto-registers itself with the Dauphin reigstry (in `dagit.dagit.schema.dauphin`) that we later lookup according to the `name` field. Spoooooky 👻! Example below...
```
Comment (schrockn) [inline python_modules/dagster/docs/guides/adding_event.md]: lol

File: python_modules/dagster/docs/guides/adding_event.md [hunk 3]
```diff
@@ -0,0 +1,90 @@
+## How to add a new event for comsumption by the GraphQL API
```
Comment (schrockn) [inline python_modules/dagster/docs/guides/adding_event.md]: *Fantastic* writeup. We should probably find a better place to but internal-only docs like this but this is fine for now

File: python_modules/dagstermill/dagstermill/__init__.py [hunk 4]
```diff
@@ -78,7 +78,7 @@ def define_out_of_pipeline_info(self, context_config):
         with yield_context(
             pipeline_def, dummy_environment_config, ExecutionMetadata(run_id='')
         ) as context:
-            self.info = TransformExecutionInfo(context, None, solid, pipeline_def)
+            self.info = TransformExecutionInfo(context, None, None, pipeline_def)
```
Comment (schrockn) [inline python_modules/dagstermill/dagstermill/__init__.py]: comment-worthy as this looks scary

File: python_modules/dagstermill/dagstermill/__init__.py [hunk 5]
```diff
@@ -115,11 +115,9 @@ def populate_context(
         # do not include event_callback in ExecutionMetadata,
         # since that'll be taken care of by side-channel established by event_logger
         execution_metadata = ExecutionMetadata(run_id, loggers=loggers)
-        solid = Solid(self.solid_def_name, self.solid_def)
         typed_environment = construct_environment_config(environment_config)
         with yield_context(pipeline_def, typed_environment, execution_metadata) as context:
-            solid_config = None
-            self.info = TransformExecutionInfo(context, solid_config, solid, pipeline_def)
+            self.info = TransformExecutionInfo(context, None, None, pipeline_def)
```
Comment (schrockn) [inline python_modules/dagstermill/dagstermill/__init__.py]: same here. Passing none is concerning.

File: python_modules/dagstermill/dagstermill/__init__.py [hunk 6]
```diff
@@ -390,10 +383,10 @@ def _t_fn(info, inputs):
                                 current_time = new_time
 
             if process.returncode != 0:
-                # Throw event that is an execution error!
-                info.log.debug("There was an error in Papermill!")
-                info.log.debug('stderr was None' if process.stderr is None else process.stderr)
-                exit()
+                raise DagstermillError(
+                    "There wsas an error when Papermill tried to execute the notebook. "
```
Comment (schrockn) [inline python_modules/dagstermill/dagstermill/__init__.py]: nit: single quote plz

File: python_modules/dagstermill/docs/intro_tutorial.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+```
+a = 3
+b = 5
+```
+```
+result = a + b
+```
+Your notebook is effectively a function that takes in inputs `a` and `b` and products output `result`. To use the language of the dagster abstraction, it is a solid with inputs `a`, `b` of dagster-type `Int` and with an ouput `result` of type `Int`. 
+
+To register this notebook as a dagster solid, we use the following lines of code.
```
Comment (schrockn) [inline python_modules/dagstermill/docs/intro_tutorial.md]: I would word this as "creating a notebook-driven solid" or something.

**Comment by schrockn** [comment]
Just some general feedback: @puma314 @bengotow I'd prefer in the future to try to make feature branches like these 1) more short-lived and 2) authored by a single human. We should be merging back into master for collaborative purposes so that we don't diverge too much and get ourselves into merge hell. Also right now we default to squash-and-merge so we might kill property attribution in the git log. We may want to the merge manually. However we should rebase so that the master history is clean.

**Comment by schrockn** [comment]
oic this branched off of materialization event

File: python_modules/dagstermill/dagstermill/__init__.py (hunk 6 above)
Comment (puma314) [reply python_modules/dagstermill/dagstermill/__init__.py]: Fixed in materialization_event pull request

**Pull request closed without merging**